service/storage_service.cc

```diff
@@ -7493,20 +7495,59 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
-        co_await streamer->stream_async();
+        switch (trinfo->transition) {
+            case locator::tablet_transition_kind::migration: {
+                auto leaving_replica_ip = host2ip(leaving_replica.host);
+
+                auto& table = _db.local().find_column_family(tablet.table);
+                std::vector<sstring> tables = {table.schema()->cf_name()};
+                auto streamer = make_lw_shared<dht::range_streamer>(_db, _stream_manager, std::move(tm),
+                                                                    guard.get_abort_source(),
+                                                                    tm->get_my_id(), _snitch.local()->get_location(),
+                                                                    "Tablet migration",
+                                                                    streaming::stream_reason::tablet_migration,
+                                                                    topo_guard,
+                                                                    std::move(tables));
+                tm = nullptr;
+                streamer->add_source_filter(std::make_unique<dht::range_streamer::failure_detector_source_filter>(
+                        _gossiper.get_unreachable_members()));
+
+                std::unordered_map<inet_address, dht::token_range_vector> ranges_per_endpoint;
+                ranges_per_endpoint[leaving_replica_ip].emplace_back(range);
+                streamer->add_rx_ranges(table.schema()->ks_name(), std::move(ranges_per_endpoint));
+                co_await streamer->stream_async();
+                break;
+            }
+            case locator::tablet_transition_kind::rebuild: {
+                // The following streams data from all replicas which are not marked as excluded
+                // Into the pending replica.
+                // FIXME: Use repair
+                auto excluded = _topology_state_machine._topology.get_excluded_nodes();
+                auto& table = _db.local().find_column_family(tablet.table);
+                std::vector<sstring> tables = {table.schema()->cf_name()};
+                auto streamer = make_lw_shared<dht::range_streamer>(_db, _stream_manager, std::move(tm),
+                                                                    guard.get_abort_source(),
+                                                                    tm->get_my_id(), _snitch.local()->get_location(),
+                                                                    "Tablet rebuild",
+                                                                    streaming::stream_reason::tablet_rebuild,
+                                                                    topo_guard,
+                                                                    std::move(tables));
+                tm = nullptr;
+                streamer->add_source_filter(std::make_unique<dht::range_streamer::failure_detector_source_filter>(
```

Comment:
I think we do not need this gossip down node filter anymore. If the chosen to sync nodes are down, we should fail the request and retry again. If the node is down and the user indeed want to ignore it, the user should use ignore dead option explicitly.

Same for the existing locator::tablet_transition_kind::migration case.

Reply:
No. The filter will not be used if the new streamer->add_rx_ranges() interface is used, which provides the address to stream from directly. The filter is used for the old streamer->add_ranges() interface which selects which nodes to stream from.  We should not use the filter here.

Reply:
Ok.